File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here

def first_then_lower_case(strings, predicate, /):
    for string in strings:
        if predicate(string):
            return string.lower()
    return None


# Write your powers generator here

def powers_generator(*, base, limit):
    power = 0
    result = 1 # 0 power always yields 1
    while result <= limit:
        yield result
        power += 1
        result = base ** power

# Write your say function here

def say(word=None, /):
    if word == None:
        return ""
    def chain(next_word=None):
        if next_word == None:
            return word
        return say(f"{word} {next_word}")
    return chain


# Write your line count function here

def meaningful_line_count(filename):
    count = 0
    try:
        with open(filename, 'r') as file:
            lines = file.readlines()
            for line in lines:
                if line.strip() and not line.lstrip().startswith('#'):
                    count += 1
        return count
    except FileNotFoundError as error:
        raise error


# Write your Quaternion class here

File: haskell/Exercises.hs
module Exercises
    ( change,
      firstThenApply,
      powers,
      meaningfulLineCount,
      Shape(..),
      volume,
      surfaceArea,
      BST(Empty),
      insert,
      contains,
      size,
      inorder
    ) where

import qualified Data.Map as Map
import Data.Text (pack, unpack, replace)
import Data.List(isPrefixOf, find)
import Data.Char(isSpace)

change :: Integer -> Either String (Map.Map Integer Integer)
change amount
    | amount < 0 = Left "amount cannot be negative"
    | otherwise = Right $ changeHelper [25, 10, 5, 1] amount Map.empty
        where
          changeHelper [] remaining counts = counts
          changeHelper (d:ds) remaining counts =
            changeHelper ds newRemaining newCounts
              where
                (count, newRemaining) = remaining `divMod` d
                newCounts = Map.insert d count counts

-- Write your first then apply function here

firstThenApply :: [a] -> (a -> Bool) -> (a -> b) -> Maybe b
firstThenApply [] _ _ = Nothing
firstThenApply (x:xs) p f
    | p x = f <$> Just x
    | otherwise = firstThenApply xs p f

-- Write your infinite powers generator here

powers :: Integral a => a -> [a]
powers base = map (base^) [0..]

-- Write your line count function here

meaningfulLineCount :: FilePath -> IO Int
meaningfulLineCount path = do
    contents <- readFile path
    return $ length $ filter meaningfulLine $ lines contents
    where
        meaningfulLine line = not (all isSpace line) && not (isComment line)
        isComment line = case dropWhile isSpace line of
                          ('#':_) -> True
                          _       -> False


-- Write your shape data type here

data Shape
    = Sphere Double         
    | Box Double Double Double  
    deriving (Eq, Show)

volume :: Shape -> Double
volume (Sphere r) = (4/3) * pi * r^3
volume (Box w h d) = w * h * d

surfaceArea :: Shape -> Double
surfaceArea (Sphere r) = 4 * pi * r^2
surfaceArea (Box w h d) = 2 * (w * h + h * d + d * w)

-- Write your binary search tree algebraic type here

data BST a = Empty | Node a (BST a) (BST a) deriving Eq

insert :: Ord a => a -> BST a -> BST a
insert x Empty = Node x Empty Empty
insert x (Node y left right)
    | x < y     = Node y (insert x left) right
    | x > y     = Node y left (insert x right)
    | otherwise = Node y left right  

contains :: Ord a => a -> BST a -> Bool
contains _ Empty = False
contains x (Node y left right)
    | x < y     = contains x left
    | x > y     = contains x right
    | otherwise = True

size :: BST a -> Int
size Empty = 0
size (Node _ left right) = 1 + size left + size right

inorder :: BST a -> [a]
inorder Empty = []
inorder (Node x left right) = inorder left ++ [x] ++ inorder right

instance Show a => Show (BST a) where
    show Empty = "()"
    show (Node x Empty Empty) = "(" ++ show x ++ ")"
    show (Node x left Empty) = "(" ++ show left ++ show x ++ ")"
    show (Node x Empty right) = "(" ++ show x ++ show right ++ ")"
    show (Node x left right) = "(" ++ show left ++ show x ++ show right ++ ")"